### ISSUE TITLE
Test titles are mixed up when randomised

### ISSUE THREAD

**Issue by backspace** [trentmwillis/ember-exam#99]
This is somewhat of a successor to #86, though it’s happening just with randomisation and not parallelisation.

You can see an example [in this `travis-web` log](https://travis-ci.org/travis-ci/travis-web/jobs/245228969#L799):

```
ok 574 Chrome 59.0 - Unit | Utility | format commit: visiting a job with a complex log
```

This is actually from [this acceptance test](https://github.com/travis-ci/travis-web/blob/24423a009e1caaa9f8f9cb276c2a4e97489ff266/tests/acceptance/job/basic-layout-test.js#L67), not a unit test. Other tests from that same file appear elsewhere:

```
ok 271 Chrome 59.0 - Acceptance | job/basic layout: visiting a job with a truncated log
…
ok 401 Chrome 59.0 - computed-limit: visiting job-view
```

The tests before the initial one I showed all have the same prefix:

```
ok 570 Chrome 59.0 - Unit | Utility | format commit: integration/components/builds-item-test.js
ok 571 Chrome 59.0 - Unit | Utility | format commit: acceptance/profile/update-hooks-test.js
ok 572 Chrome 59.0 - Unit | Utility | format commit: acceptance/repo/build-list-routes-test.js
ok 573 Chrome 59.0 - Unit | Utility | format commit: serializers/repo_v2_fallback.js
```

I know that [you said](https://github.com/trentmwillis/ember-exam/issues/86#issuecomment-299211024) that this is probably a QUnit issue but I’m not sure how to actually prove that. If you have any suggestions on how I can show that it’s a QUnit problem and not Ember Exam, let me know, and I can open an issue there instead!

Since this is a public repository, you can reproduce this phenomenon locally with `ember exam --random=yq9pfker5xi0xnp0f2kfhia4i`. If you’re interested in that and have any trouble getting it running, I can make some suggestions.

Thanks for your work on this, very useful!

**Comment by mani-mishra** [comment]
Observing this in our repo as well. 
For us, there are some tests which are failing too, w/o any apparent dependency or leakage, when running with default seed. But if I provide a custom seed value, those tests work. :man_shrugging: 

I was wondering if its relate to this mix up or not.

@backspace do you see any random test failures as well?

**Comment by backspace** [comment]
While the `travis-web` test suite does show occasional intermittent failures, it’s my belief that they’re not related to this, because they don’t reproduce reliably even with the same test seed. The one I can think of is because of some questionable use of `setTimeout` 😳

**Comment by trentmwillis** [comment]
Sorry for not replying to this sooner. Since this issue is only with the `random` option, it will definitely be a problem within QUnit itself and not this addon. This addon simply [sets a query param](https://github.com/trentmwillis/ember-exam/blob/dedef34b6677ca618bba8078cd1c96df1a37359c/lib/commands/exam.js#L116) that is placed into the test page URL which triggers QUnit's randomization logic.

That said, I'll leave this issue open and begin exploring what might be happening. Thanks for the reproduction as well, should make it a lot easier.

**Comment by trentmwillis** [comment]
Actually, after looking at the logs, it struck me that this is likely just a reporter bug. I believe the issue lies in Testem's `qunit_adapter`, which [manually tracks the current module being run](https://github.com/testem/testem/blob/4e6e484ddc687d14c9274b84855a94ab7795a365/public/testem/qunit_adapter.js#L112) instead of using the module name as reported with the logging info.

**Comment by trentmwillis** [comment]
I've opened a PR to fix this in Testem: https://github.com/testem/testem/pull/1155. I've verified locally that it addresses the problem.

Now, for those that wish to understand why this started occurring. Previously, QUnit used to call `moduleStart` and `moduleEnd` callbacks any time the current module changed. In a non-randomized world this works fine, but with random tests you wind up getting many more `moduleStart` and `moduleEnd` triggers than you actually need.

So, we treated this as a bug fix, and changed the behavior to only ever fire these once per module. `moduleStart` occurs before the first test in the module and `moduleEnd` occurs only after all tests have finished running. However, this broke the manual tracking of the current module that Testem was doing. Hopefully with that context, the fix linked above makes sense.
